cmd/segment-reaper/observer.go

```diff
@@ -143,11 +145,88 @@ func (obsvr *observer) processSegment(ctx context.Context, path metainfo.ScopedP
 // analyzeProject analyzes the objects in obsv.objects field for detecting bad
 // segments and writing them to objs.writer.
 func (obsvr *observer) analyzeProject(ctx context.Context) error {
-	// TODO this part will be implemented in next PR
-	// TODO(if): For what is this?
+	for bucket, objects := range obsvr.objects {
+		for path, object := range objects {
+			if object.skip {
+				continue
+			}
+
+			brokenObject := false
+			if object.hasLastSegment {
+				if object.expectedNumberOfSegments == 0 {
+					if !object.segments.IsSequence() {
+						brokenObject = true
+					}
+				} else if object.segments.Count() != int(object.expectedNumberOfSegments)-1 {
+					// TODO should we also check if its valid sequence
+					// expectedNumberOfSegments-1 because 'segments' doesn't contain last segment
+					brokenObject = true
+				}
+
+				if brokenObject {
+					err := obsvr.printSegment(ctx, "l", bucket, path, obsvr.writer)
+					if err != nil {
+						return err
+					}
+				}
+			} else {
+				brokenObject = true
+			}
+
+			if brokenObject {
+				for index := 0; index < maxNumOfSegments; index++ {
+					has, err := object.segments.Has(index)
+					if err != nil {
+						return nil
+					}
+					if has {
+						err := obsvr.printSegment(ctx, "s"+strconv.Itoa(index), bucket, path, obsvr.writer)
+						if err != nil {
+							return err
+						}
+					}
+				}
+			}
+		}
+	}
 	return nil
 }
 
+func (obsvr *observer) printSegment(ctx context.Context, bucket string, segmentIndex string, path string, csvWriter *csv.Writer) error {
```

Comment:
any reason not to combine the `string`s?

Reply:
internally this method needs these values separated.

Reply:
I just meant to use a single type declaration for `bucket`, `segmentIndex` and `path`.

Reply:
I know but for example `path` needs to be separate parameter because it needs to be encoded so I don't think make sense to combine only segmentIndex and bucket

Reply:
I mean make the signature this
```
 func (obsvr *observer) printSegment(ctx context.Context, bucket, segmentIndex, path string, csvWriter *csv.Writer) error
```

Reply:
ahh, yeah good idea:) I need to warm up after a long weekend.